mdetsims/metacal/metacal_fitter.py

```diff
@@ -0,0 +1,284 @@
+
+            # this gets all objects, all bands in a list of MultiBandObsList
+            mbobs_list_mcal = fitter.make_corrected_obs()
+        else:
+            mbobs_list_mcal = mbobs_list
+            mof_data = None
+        self.mbobs_list_mcal = mbobs_list_mcal
+
+        if self['metacal']['symmetrize_weight']:
+            self._symmetrize_weights(mbobs_list_mcal)
+
+        self._result = self._do_all_metacal(mbobs_list_mcal, data=mof_data)
+
+    def _symmetrize_weights(self, mbobs_list):
+        def _symmetrize_weight(wt):
+            assert wt.shape[0] == wt.shape[1]
+            wt_rot = np.rot90(wt)
+            wzero = wt_rot == 0.0
+            if np.any(wzero):
+                wt[wzero] = 0.0
+
+        for mbobs in mbobs_list:
+            for obslist in mbobs:
+                for obs in obslist:
+                    _symmetrize_weight(obs.weight)
+
+    def _do_all_metacal(self, mbobs_list, data=None):
+        """run metacal on all objects
+
+        NOTE: failed mbobs will have no entries in the final list
+        """
+
+        assert len(mbobs_list[0]) == self.nband
+
+        datalist = []
+        for i, mbobs in enumerate(mbobs_list):
+            if self._check_flags(mbobs):
+                try:
+                    boot = self._do_one_metacal(mbobs)
+                    if isinstance(boot, dict):
+                        res = boot
+                    else:
+                        res = boot.get_metacal_result()
+                except (BootPSFFailure, BootGalFailure) as err:
+                    logger.debug(str(err))
+                    res = {'mcal_flags': 1}
+
+                if res['mcal_flags'] != 0:
+                    logger.debug("metacal fit failed")
+                else:
+                    # make sure we send an array
+                    fit_data = self._get_metacal_output(res, self.nband, mbobs)
+                    if data is not None:
+                        odata = data[i:i+1]
+                        fit_data = eu.numpy_util.add_fields(
+                            fit_data,
+                            odata.dtype.descr,
+                        )
+                        eu.numpy_util.copy_fields(odata, fit_data)
+
+                    self._print_result(fit_data)
+                    datalist.append(fit_data)
+
+        if len(datalist) == 0:
+            return None
+
+        output = eu.numpy_util.combine_arrlist(datalist)
+        return output
+
+    def _do_one_metacal(self, mbobs):
+        conf = self['metacal']
+        psf_pars = conf['psf']
+        max_conf = conf['max_pars']
+
+        tpsf_obs = mbobs[0][0].psf
+        if not tpsf_obs.has_gmix():
+            _fit_one_psf(tpsf_obs, psf_pars)
+
+        psf_Tguess = tpsf_obs.gmix.get_T()
+
+        boot = self._get_bootstrapper(mbobs)
+        if 'lm_pars' in psf_pars:
+            psf_fit_pars = psf_pars['lm_pars']
+        else:
+            psf_fit_pars = None
+
+        prior = self.metacal_prior
```

Comment:
I now assume we always have an input prior for metacal.

Reply:
good; this is fine for max metacal (not used for e.g. adaptive moments)